mat64/shadow.go

```diff
@@ -223,3 +220,62 @@ func rectanglesOverlap(off, aCols, bCols, stride int) bool {
 	// b strictly wraps and so must overlap with a.
 	return true
 }
+
+// trianglesOverlap returns whether the strided triangles a and b overlap
+// when b is offset by off elements after a but has at least one element before
+// the end of a. a and b have aSize and bSize respectively.
+func trianglesOverlap(off, aSize, bSize, stride int, aUpper, bUpper bool) bool {
+	if !rectanglesOverlap(off, aSize, bSize, stride) {
+		// Fast return if bounding rectangles do not overlap.
+		return false
+	}
+
+	if aUpper {
+		// Test overlap by projecting diagonals from the b upper
+		// left corner, and upper right or lower left corner onto
+		// the top row of a.
+		rows := off / stride
+		proj := off - rows*(stride+1)
+		if proj >= 0 && proj < aSize {
```

Comment:
If there is no `rectanglesOverlap` then this condition will be true in this situation:
```
aaa.
.aab
..a.
```
because
`aSize = 3`, `stride = 4`, `off = 7`.
Then
`rows = 1`, `proj = 7 - 1*5 = 2`.
And `2 >= 0 && 2 < 3` is true.
But `proj` in fact is the distance from a's diagonal. So if the condition is changed to
`if proj >= 0 && proj < aSize - rows`
it will be correctly false even without rectanglesOverlap.

Maybe something similar would work also in the !aUpper case but I haven't considered it.

I hope I'm not complicating the matter unnecessarily but for me it's a good way to understand this better. And getting the notion of projection to the top row away would simplify the reasoning for me.